hazelcast/src/main/java/com/hazelcast/internal/serialization/impl/GenericRecordQueryReader.java

```diff
@@ -0,0 +1,322 @@
+            // ex: attribute any
+            while (iterator.hasNext()) {
+                InternalGenericRecord record = (InternalGenericRecord) iterator.next();
+                iterator.remove();
+                Object leaves = readLeaf(record, fieldName);
+                if (leaves == null) {
+                    multiResult.setNullOrEmptyTarget(true);
+                } else if (leaves instanceof Object[]) {
+                    Object[] array = (Object[]) leaves;
+                    if (array.length == 0) {
+                        multiResult.setNullOrEmptyTarget(true);
+                        continue;
+                    }
+                    for (Object leaf : array) {
+                        iterator.add(leaf);
+                    }
+                } else {
+                    assert leaves.getClass().isArray() : "parameter is not an array";
+                    if (!ExtractorHelper.reducePrimitiveArrayInto(iterator::add, leaves)) {
+                        multiResult.setNullOrEmptyTarget(true);
+                    }
+                }
+            }
+        } else {
+            // ex: attribute[2]
+            int index = Integer.parseInt(extractArgumentsFromAttributeName(path));
+            while (iterator.hasNext()) {
+                GenericRecord record = (GenericRecord) iterator.next();
+                Object leaf = readIndexed((InternalGenericRecord) record, fieldName, index);
+                iterator.set(leaf);
+            }
+        }
+
+        if (multiResult.isNullEmptyTarget()) {
+            results.addFirst(null);
+        } else if (results.size() == 1) {
+            return results.get(0);
+        }
+        return multiResult;
+    }
+
+    private <T> T readIndexed(InternalGenericRecord record, String path, int index) throws IOException {
+        if (!record.hasField(path)) {
+            return null;
+        }
+        FieldType type = record.getFieldType(path);
+        switch (type) {
+            case BYTE_ARRAY:
+                return (T) record.readByteFromArray(path, index);
+            case SHORT_ARRAY:
+                return (T) record.readShortFromArray(path, index);
+            case INT_ARRAY:
+                return (T) record.readIntFromArray(path, index);
+            case LONG_ARRAY:
+                return (T) record.readLongFromArray(path, index);
+            case FLOAT_ARRAY:
+                return (T) record.readFloatFromArray(path, index);
+            case DOUBLE_ARRAY:
+                return (T) record.readDoubleFromArray(path, index);
+            case BOOLEAN_ARRAY:
+                return (T) record.readBooleanFromArray(path, index);
+            case CHAR_ARRAY:
+                return (T) record.readCharFromArray(path, index);
+            case UTF_ARRAY:
+                return (T) record.readUTFFromArray(path, index);
+            case PORTABLE_ARRAY:
+                return (T) record.readObjectFromArray(path, index);
```

Comment:
In `readLeaf` and `readIndexed` T can never be generic record. Having said that, Generics does not seem useful here at all because I am always assigning to Object. I will replace Generics with Object.